### ISSUE TITLE
Request: Return old default functionality from 2.1.1 to now

### ISSUE THREAD

**Issue by AndreLouisIssa** [EastDesire/jscolor#21]
The most recent versions (later than 18th Jun) do not work on my userscripts anymore, I had to try to work around this but it never ended up working the same.

Please return it to how it deploys in the old version or maintain this function on a separate branch.
For now I have changed the require to use a specific version from the old commit in order to keep things working.

How does it not work?:
The input with class jscolor no longer does anything but show ffffff instead of FFFFFF and doesn't bring up a picker at all
then if I use the apparently new method of initialising the thing by doing inputElement.jscolor = new jscolor(inputElement)
it works inside the element but doesn't look right and it also doesn't deploy the onFineChange anymore

I don't know any more specifics, it just doesn't work, and working around it comes with so many drawbacks I was only able to do it successfuly for one of my projects

Project that fails to work now and I can't rework the new version to make it work: https://github.com/MagicGonads/MagicGonads.github.io/blob/master/SketchfulPicker.js

Project that fails to work now but I was able to rework the new version to make it work:
https://github.com/AEFGP/AEFGP.github.io/blob/master/Userscripts/Desmos/ColourPicker.js

**Comment by EastDesire** [comment]
Hi @MagicGonads ,

SketchulPicker.js:
I tried the following code of yours and it worked with the latest version:

```js
    let nodePicker = document.createElement ('input');
    nodePicker.setAttribute ('id', 'picker');
    nodePicker.setAttribute ('class', `jscolor {onFineChange:"document.getElementById('picker').oninput()"}`);
    nodePicker.setAttribute ('value',"ffffff");
    nodePicker.oninput = function() {
        swapColor(0);
    }
    nodePicker.setAttribute ('style',"position: relative; text-align: center; left: 1%; width:30%; top:-30%; height:30%; border:0px solid currentcolor");
    anchor.appendChild(nodePicker);
```

However:
- installing jscolor by class name is now deprecated (you should see a warning in the console when using latest version). Instead, use **data-jscolor** attribute
- **onFineChange** is now replaced by **onInput**
- the standard oninput handler is now triggered automatically on the input element, so it is not necessary to explicitly trigger oninput()

In short, taking the above into account, you can replace this line:
```js
    nodePicker.setAttribute ('class', `jscolor {onFineChange:"document.getElementById('picker').oninput()"}`);
```
with this:
```js
    nodePicker.setAttribute ('data-jscolor', '{}');
```

PS: It would be an overkill to maintain the old functionality if the last version can work fine when used properly.

If it still won't work for you, please make an online demo so that I can test it in a browser. The above code worked as expected for me with the latest jscolor version - it showed the picker and tried to call swapColor(0); during color changes.

**Comment by EastDesire** [comment]
> then if I use the apparently new method of initialising the thing by doing inputElement.jscolor = new jscolor(inputElement)
> it works inside the element but **doesn't look right** and it also doesn't deploy the onFineChange anymore

If you mean that the field has a color preview on its left side instead of covering the entire element - yes, this is intended. Because the new version also supports alpha channel and the text would't be readable with transparency chessboard behind it (check [jscolor examples](https://jscolor.com/examples/) to see what I mean).

**Comment by EastDesire** [comment]
> then if I use the apparently new method of initialising the thing by doing inputElement.jscolor = new jscolor(inputElement)
> it works inside the element but **doesn't look right** and it also doesn't deploy the onFineChange anymore

If you mean that the field has a color preview on its left side instead of covering the entire element - yes, this is intended. Because the new version also supports alpha channel and the text wouldn't be readable with transparency chessboard behind it (check [jscolor examples](https://jscolor.com/examples/) to see what I mean).

**Comment by AndreLouisIssa** [comment]
I'm sorry but the code you sent me doesn't work, it still only works with that specific commit version and the original class with the onFineChange given

**Comment by AndreLouisIssa** [comment]
Maybe it's a result of the way tampermonkey does things, but, setting the attribute just creates a bubble with no actual picker spawning when you click it whereas before it would set it all up properly

**Comment by EastDesire** [comment]
> I'm sorry but the code you sent me doesn't work, it still only works with that specific commit version and the original class with the onFineChange given

Are there any messages in browser's developer console? What browser version are you using?

There were many changes between 2.1.1 and 2.2.0, one of them was removal of VML, but that is replaced by Canvas API and should work in modern browsers.

For me to troubleshoot it, I need some kind of demo, a runnable snippet, otherwise I'll be just guessing what the issue might be.

**Comment by AndreLouisIssa** [comment]
I am using Chrome 84.0.4147.89, Tampermonkey 4.10
There are no errors in the console besides ones already produced by Sketchful.io without the script running

**Comment by AndreLouisIssa** [comment]
To test the script, visit https://sketchful.io/ then click "FREE DRAW" (yellow button near the middle of the window) (sometimes the site bugs out and sends you to a public lobby, if that happens reload and try to enter free draw again)
If (with the script enabled) the JScolor box appears to the right. When broken it displays "ffffff" (and when clicked does not spawn the picker) when working it displays "FFFFFF" (and does spawn the picker)

**Comment by AndreLouisIssa** [comment]
The specific version of jscolor I've been testing it with when it works is 
https://raw.githubusercontent.com/EastDesire/jscolor/8dff1faa16989806696a624204f6ebbc962425a7/jscolor.js
(where the latest, is obviously
https://raw.githubusercontent.com/EastDesire/jscolor/master/jscolor.js
)

**Comment by EastDesire** [comment]
Is it possible that Tampermonkey doesn't trigger DOMContentLoaded event in its scripts? This is probably why the new versions don't work.

To make your script work with the latest jscolor and also emulate the old way of displaying color preview in the input:

1. Replace line 78:
```js
    nodePicker.setAttribute ('class', `jscolor {onFineChange:"document.getElementById('picker').oninput()"}`);
```
with this:
```js
    nodePicker.setAttribute ('data-jscolor', '{hash:false, previewElement:null}');
```

2. Update oninput callback to this:
```js
    nodePicker.oninput = function () {
        // emulate old jscolor look
        this.style.backgroundColor = this.jscolor.toHEXString();
        this.style.color = this.jscolor.isLight() ? 'black' : 'white';
        // use chosen color
        swapColor(0);
    }
```

3. Add the following line after `anchor.appendChild(nodePicker);` or somewhere at the end of the script:
```js
    jscolor.install(); // we need to install jscolor explicitly
```

4. Recommendation: Wherever you use `nodePicker.style.backgroundColor`, replace it with `nodePicker.jscolor.toHEXString()` (I think it's used at 2 places)

**Comment by AndreLouisIssa** [comment]
That did work, thanks.
It does seem to be something with Tampermonkey then

**Comment by EastDesire** [comment]
> That did work, thanks.

Good, you're welcome.

> It does seem to be something with Tampermonkey then

I'd guess this is an intended behavior of Tampermonkey. But old jscolor versions also installed on window.onload etc., which was later removed. So I guess the old versions **happened to work** in Tampermonkey without calling jscolor.install().
